WIP: Added the '-l' option for rahash2

ITAYC0HEN opened the pull request:
Rebased and up-to-date version of PR #8678 (now closed)

## Updates:
**binr/rahash2/rahash2.c**

* Added the functionality to not output the trailing newline. Until now, some of the algorithms outputted newlines (e.g. md5) and some was not (e.g base64, rot13). Now the flag "-l" is responsible for not printing newline.


**man/rahash2.1**

* Added to the manpage the flag "-l" for not trailing newline.
* Added the missing entry for "-n".

radare commented:
looks red

ITAYC0HEN commented:
Fixed tests: https://github.com/radare/radare2-regressions/pull/1357

xarkes commented:
FYI I think that if you name your r2 regressions branch the same as your r2 branch, travis will fetch your regressions to test the PR.

ITAYC0HEN commented:
Oh nice! Didn't know that 😱😎

On Mon, Jun 18, 2018, 10:08 xarkes <notifications@github.com> wrote:

> FYI I think that if you name your r2 regressions branch the same as your
> r2 branch, travis will fetch your regressions to test the PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/radare/radare2/pull/10341#issuecomment-397961190>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATP2ch2RUGtt3EppZv1CPQaK8Zm4aqpaks5t91HvgaJpZM4UmpKB>
> .
>

ITAYC0HEN commented:
Closing and opening to trigger Travis

radare commented:
ping

ITAYC0HEN commented:
I tried to solve it for hours and still, struggling with making Travis green :(
I've already fixed the vast majority of issues (https://github.com/radare/radare2-regressions/pull/1368).

Would really appreciate any help with fixing the rest of the regressions so we can merge this PR finally.

radare commented:
please squash and rebase

radare commented:
if you dont do that travis will never be green

ITAYC0HEN commented:
Rebased, I think:
![image](https://user-images.githubusercontent.com/20182642/42731565-63a8d6ec-8818-11e8-973e-5e2ba5485190.png)

still red ;(

radare commented on binr/rahash2/rahash2.c:
```diff
@@ -23,9 +24,9 @@ void compare_hashes(const RHash *ctx, const ut8 *compare, int length, int *ret)
 		// algobit has only 1 bit set
 		if (!memcmp (ctx->digest, compare, length)) {
-			printf ("rahash2: Computed hash matches the expected one.\n");
+			printf ("\nrahash2: Computed hash matches the expected one.\n");
```
why this newline

ITAYC0HEN replied:
Because now, without `-l` there won't be a new line:

```
// No new line
Megabbets-PC$ rahash2 -a crc32 -s test
0x00000000-0x00000003 crc32: d87f7e0cMegabbets-PC$

// with new line
Megabbets-PC$ rahash2 -l -a crc32 -s test
0x00000000-0x00000003 crc32: d87f7e0c

// \n is required in order to print "Computed hash matches the expected one." in a new line
Megabbets-PC$ rahash2 -a crc32 -s test -c d87f7e0c
0x00000000-0x00000003 crc32: d87f7e0c
rahash2: Computed hash matches the expected one.
```

On master you have new line after the result of the hash, now you won't (without `-l`)

radare commented on binr/rahash2/rahash2.c:
```diff
@@ -23,9 +24,9 @@ void compare_hashes(const RHash *ctx, const ut8 *compare, int length, int *ret)
+			printf ("\nrahash2: Computed hash matches the expected one.\n");
 		} else {
-			eprintf ("rahash2: Computed hash doesn't match the expected one.\n");
+			eprintf ("\nrahash2: Computed hash doesn't match the expected one.\n");
```
why this newline

ITAYC0HEN replied:
same as above

radare commented on binr/rahash2/rahash2.c:
```diff
@@ -216,9 +221,10 @@ static int do_hash(const char *file, const char *algo, RIO *io, int bsize, int r
+					printf (" %s", file);
+				} 
+				if (linebreak) {
+					if (quiet==1 || (quiet && !rad)) {
```
spaces around ==

radare commented on binr/rahash2/rahash2.c:
```diff
@@ -356,6 +364,10 @@ static int encrypt_or_decrypt(const char *algo, int direction, const char *hashs
 					write (1, result, result_size);
+					// Add linebreak by default
+					if (linebreak) {
+						write (1,  "\n", 1);
```
adding a line break by default in the output of an encrypted or decrypted output will break the contents and make it unusable

radare commented:
all the XX in travis are because your branch is not rebased properly. please squash all your commits into one and rebase it on top of master

radare commented:
and clsing/opening doesnt rebases anything

xarkes commented:
Yes just `git rebase master` and fix conflicts if any.

ITAYC0HEN commented:
Squashed and Rebased ( I hope )
"This branch is 1 commit ahead of radare:master."
"ITAYC0HEN wants to merge 1 commit into radare:master from ITAYC0HEN:fixNewlineRahash2"

radare commented on binr/rahash2/rahash2.c:
```diff
@@ -15,6 +15,7 @@ static ut64 to = 0LL;
 static int incremental = 1;
 static int iterations = 0;
 static int quiet = 0;
+static int linebreak = 0;
```
use bool instead of int, and false/true, same goes for `quiet`

XVilka commented:
@ITAYC0HEN please rebase

ITAYC0HEN commented:
I need to change the functionality again as agreed with pancake. will rebase again as soon as I get to this :{
Sorry, uber busy recently (r2 stuff mostly)

radare commented:
would be nice before r2con. can you rebase?

radare commented:
Nooooooooo never merge

ret2libc commented:
Is this PR really necessary?
It seems to me rahash2 is quite consistent (unless I'm missing something), because all hasing algorithms (the one you specify with `-a`) print a newline, while all the encoders/"encryptors" (the one with `-E`/`-D`) do not. I guess this is done to make it easy to redirect the output to file.

radare commented:
The only reason for this is to appwnd a newline when decoding base64 strings (or decrypting) that end without newlines which hide the output under the prompt

> On 31 Aug 2018, at 10:34, Riccardo Schirone <notifications@github.com> wrote:
> 
> Is this PR really necessary?
> It seems to me rahash2 is quite consistent (unless I'm missing something), because all hasing algorithms (the one you specify with -a) print a newline, while all the encoders/"encryptors" (the one with -E/-D) do not. I guess this is done to make it easy to redirect the output to file.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

ret2libc commented:
I see. But can't we just make the `r_crypto_plugin_base64` print a newline in the `final` method? I don't think it makes much sense for the other crypto algorithms to add a newline (it may even "corrupt" the encrypted otuput, i guess).

radare commented:
Well i want it to be consistent by default and others want to be user friendly. Thats the thing to decide

> On 31 Aug 2018, at 13:39, Riccardo Schirone <notifications@github.com> wrote:
> 
> I see. But can't we just make the r_crypto_plugin_base64 print a newline in the final method? I don't think it makes much sense for the other crypto algorithms to add a newline (it may even "corrupt" the encrypted otuput, i guess).
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

ret2libc commented:
The fact is that if you add a newline to the encrypted output of some algorithm, you are going to "corrupt" the encrypted string. If you add a newline to base64, then everything is still good.

If you do `rahash2 -E xor -S ffffff /bin/ls` you are going to get a lot of garbage and it doesn't make sense to add a newline at the end. In case of base64 (and maybe others) it makes sense because the output is still ascii.

I'd vote for just adding a newline on base64, unless there are many other cases that require the newline.

radare commented:
> On 31 Aug 2018, at 14:15, Riccardo Schirone <notifications@github.com> wrote:
> 
> The fact is that if you add a newline to the encrypted output of some algorithm, you are going to "corrupt" the encrypted string. If you add a newline to base64, then everything is still good.
> 

its good for a human, but if you reencode and decode b64 by adding a newline everytime you’ll end with many newlines
> If you do rahash2 -E xor -S ffffff /bin/ls you are going to get a lot of garbage and it doesn't make sense to add a newline at the end. In case of base64 (and maybe others) it makes sense because the output is still ascii
> 
> 
i use to solve this by appending ‘; echo’ at the end of the command to ensure a newline
> I'd vote for just adding a newline on base64, unless there are many other cases that require the newline.
> 

also this pr fixes the manpage, which was missing one flag

radare commented:
I think this pr got stuck for enough time, and having a newline at the end of the textual representation of a hash shouldnt be a problem, remving it causes more dirt than i expected in the source.. so closing.
